ShaiRandom/GapShuffler.cs

```diff
@@ -0,0 +1,283 @@
+    /// array, as well a a similar constructor which copies a Span.  In the case of in-place enumeration, like most
+    /// enumerators, modifying the collection while the shuffle is active (including adding or removing items) will cause
+    /// undefined behavior.
+    ///
+    /// This type is a struct, and as such is much more efficient when used in a foreach loop than a function returning
+    /// IEnumerable&lt;TItem&gt; by using "yield return".  This type does implement <see cref="IEnumerable{TItem}"/>,
+    /// so you can pass it to functions which require one (for example, System.LINQ).  However, this will have reduced
+    /// performance due to boxing of the iterator.
+    /// </remarks>
+    public struct GapShufflerEnumerator<TItem> : IEnumerator<TItem>, IEnumerable<TItem>
+    {
+        private readonly IList<TItem> _items;
+        private readonly IEnhancedRandom _rng;
+        private int _index;
+
+        // We store the size instead of using _items.Count in MoveNext because changing the items in the middle of
+        // an enumeration is not useful behavior in either case, and this is faster for many collections.
+        private readonly int _size;
+
+        // Suppress warning stating to use auto-property because we want to guarantee micro-performance
+        // characteristics.
+#pragma warning disable IDE0032 // Use auto property
+        private TItem _current;
+#pragma warning restore IDE0032 // Use auto property
+
+        /// <summary>
+        /// The current value for enumeration.
+        /// </summary>
+        public TItem Current => _current;
+
+        /// <inheritdoc />
+        object? IEnumerator.Current => _current;
+
+        /// <summary>
+        /// Creates an enumerator that implements the "gap shuffle" algorithm.  This constructor copies the IEnumerable
+        /// of items it receives into an array, which it uses internally for the shuffle.
+        /// </summary>
+        /// <param name="rng">RNG to use for shuffling.</param>
+        /// <param name="items">Items to shuffle, which are copied into an array internally.</param>
+        public GapShufflerEnumerator(IEnhancedRandom rng, IEnumerable<TItem> items)
+        {
+            _items = items.ToArray();
+            _size = _items.Count;
+            _rng = rng;
+
+            rng.Shuffle(_items);
+
+            _index = 0;
+            _current = default!;
+        }
+
+        /// <summary>
+        /// Creates an enumerator that implements the "gap shuffle" algorithm.  This constructor copies the span
+        /// of items it receives into an array, which it uses internally for the shuffle.
+        /// </summary>
+        /// <param name="rng">RNG to use for shuffling.</param>
+        /// <param name="items">Items to shuffle, which are copied into an array internally.</param>
+        public GapShufflerEnumerator(IEnhancedRandom rng, ReadOnlySpan<TItem> items)
```

Comment:
It's a little funny that this can mutate a `ReadOnlySpan<TItem>`, though I'm not complaining. I guess they aren't really read-only!

Reply:
It doesn't mutate it, it copies it into an array; so it never mutates the span, it mutates the copy :D.  The reason I didn't use `Span` here is because, I _can't_ store a Span inside a struct; it's a ref struct, which functionally means it's allowed to exist on the stack and that's about it; can't use them as fields, etc; so there is no nice way to implement an in-place version of this on a Span.   Really this is what Memory<T> is for, thus the other iterator that uses `Memory<T>`